bluemix/terminal/table.go

```diff
@@ -69,58 +75,152 @@ func (t *PrintableTable) Add(row ...string) {
 	}
 }
 
+func isWideColumn(col string) bool {
+	// list of common columns that are usually wide
+	largeColumnTypes := []string{T("ID"), T("Description")}
+
+	for _, largeColn := range largeColumnTypes {
+		if strings.Contains(largeColn, col) {
+			return true
+		}
+	}
+
+	return false
+
+}
+
+func terminalWidth() int {
+	var err error
+	terminalWidth, _, err := term.GetSize(int(os.Stdin.Fd()))
+
+	if err != nil {
+		// Assume normal 80 char width line
+		terminalWidth = 80
+	}
+	return terminalWidth
+}
+
 func (t *PrintableTable) Print() {
 	for _, row := range append(t.rows, t.headers) {
 		t.calculateMaxSize(row)
 	}
 
-	if t.headerPrinted == false {
-		t.printHeader()
-		t.headerPrinted = true
+	tbl := table.NewWriter()
+	tbl.SetOutputMirror(t.writer)
+	tbl.SuppressTrailingSpaces()
+	// remove padding from the left to keep the table aligned to the left
+	tbl.Style().Box.PaddingLeft = ""
+	tbl.Style().Box.PaddingRight = strings.Repeat(" ", minSpace)
+	// remove all border and column and row separators
+	tbl.Style().Options.DrawBorder = false
+	tbl.Style().Options.SeparateColumns = false
+	tbl.Style().Options.SeparateFooter = false
+	tbl.Style().Options.SeparateHeader = false
+	tbl.Style().Options.SeparateRows = false
+	tbl.Style().Format.Header = text.FormatDefault
+
+	headerRow, rows := t.createPrettyRowsAndHeaders()
+	columnConfig := t.createColumnConfigs()
+
+	tbl.SetColumnConfigs(columnConfig)
+	tbl.AppendHeader(headerRow)
+	tbl.AppendRows(rows)
+	tbl.Render()
+}
+
+func (t *PrintableTable) createColumnConfigs() []table.ColumnConfig {
+	// there must be at row in order to configure column
+	if len(t.rows) == 0 {
+		return []table.ColumnConfig{}
+	}
+
+	colCount := len(t.rows[0])
+	var (
+		widestColIndicies []int
+		terminalWidth     = terminalWidth()
+		// total amount padding space that a row will take up
+		totalPaddingSpace = (colCount - 1) * minSpace
+		remainingSpace    = terminalWidth - totalPaddingSpace
+		// the estimated max column width by dividing the remaining space evenly across the columns
+		maxColWidth = (terminalWidth - totalPaddingSpace) / colCount
+	)
+	columnConfig := make([]table.ColumnConfig, len(t.maxSizes))
+
+	for i := range columnConfig {
+		columnConfig[i] = table.ColumnConfig{
+			AlignHeader: text.AlignLeft,
+			Align:       text.AlignLeft,
+			WidthMax:    maxColWidth,
+			Number:      i + 1,
+		}
+
+		// assuming the table has headers: store columns with wide content where the max width may need to be adjusted
+		// using the remaining space
+		if t.maxSizes[i] > maxColWidth && (len(t.headers) > 0 && isWideColumn(t.headers[i])) {
+			widestColIndicies = append(widestColIndicies, i)
+		} else if t.maxSizes[i] < maxColWidth {
+			// use the max column width instead of the estimated max column width
+			// if it is shorter
+			columnConfig[i].WidthMax = t.maxSizes[i]
+			remainingSpace -= t.maxSizes[i]
+		} else {
+			remainingSpace -= maxColWidth
```

Comment:
Wondering if there can be a scenario that remainingSpace becomes 0 or less.

Reply:
Good question. I don't think there is a valid scenario since that would mean the user would need to have a lot of columns. For instance, for terminal width of 80, the user would need to create more than 27 columns (27 * 3 > 80) for there to be negative `remainingSpace`.